### PR TITLE
Feature/33860 run ruby tests in docker

### PR DESCRIPTION
**Pull request by danwilliams**

### CODE REVIEW

File: pupil-spa/src/app/login-failure/login-failure.component.html [hunk 1]
```diff
@@ -10,7 +10,7 @@ <h1 tabindex="0">Sign-in error</h1>
         <strong class="visually-hidden">Warning</strong>
         <app-svg-warning></app-svg-warning>
         <strong>
-          Please tell your teacher
+          Please tell your teacher: {{errorMessage}}
```
Comment (GuyHarwood) [inline pupil-spa/src/app/login-failure/login-failure.component.html]: Should this be in this PR?

File: admin/Dockerfile [hunk 2]
```diff
@@ -14,4 +14,7 @@ RUN ./node_modules/.bin/gulp build
 RUN yarn global add pm2
 
 EXPOSE 3001
-CMD ["./docker-start.sh" ]
+
+RUN chmod +x wait-sqldb.sh
+
+CMD ./wait-sqldb.sh && ./docker-start.sh
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/Dockerfile]: however - cross checking config and the wait file, this should potentially work in azure anyway?  Then the question is whether thats the behaviour we want

[the remark this answers is not in the document]
Comment (danwilliams) [reply admin/Dockerfile]: Changed `wait-sqldb` to only run if an env var is provided (which I’ve added to the docker-compose for testing)

File: admin/docker-start.sh [hunk 3]
```diff
@@ -3,11 +3,8 @@
 echo "migrate is $MIGRATE."
 if [ "$MIGRATE" == "1" ]
 then
-  WAIT_TIME="20s"
-  echo "waiting $WAIT_TIME seconds for SQL Server to come online..."
-  sleep $WAIT_TIME
   echo "running database migrations..."
-  yarn migrate-sql
+  yarn migrate-and-seed-sql
```
[the remark this answers is not in the document]
Comment (danwilliams) [reply admin/docker-start.sh]: Running migrations was already there - only done if the `MIGRATE` env var is specified.
I’ve now split out seeding to work in the same way.

File: docker-compose.pupil-test.yml [hunk 4]
```diff
@@ -43,34 +62,24 @@ services:
...
-    volumes:
-    - ./test-output/admin/screenshots:/usr/src/app/screenshots
-    - ./test-output/admin/reports:/usr/src/app/reports
-    links:
-    - sqldb
-    - admin-app
     environment:
-      BASE_URL: "http://admin-app:3001"
-      SQL_SERVER: "sqldb"
+      APPINSIGHTS_INSTRUMENTATIONKEY: ${AZURE_STORAGE_CONNECTION_STRING}
```
[the remark this answers is not in the document]
Comment (danwilliams) [reply docker-compose.pupil-test.yml]: The pupil SPA needs the env var `APPINSIGHTS_INSTRUMENTATIONKEY` specified. I noticed that it happens to be the same value as `AZURE_STORAGE_CONNECTION_STRING`, which was already being used for the testing. So used it to reduce dependencies to get testing running.

File: functions/Dockerfile [hunk 5]
```diff
@@ -1,11 +1,11 @@
 ### STAGE 1: Build ###
-FROM node:10.14.1 as builder
+FROM node:10.14.1
```
[the remark this answers is not in the document]
Comment (danwilliams) [reply functions/Dockerfile]: I needed to be able to run a bash script to wait for the SQL server container to be running, which I couldn’t get to work with the official image. I’ve restored the original `functions/Dockerfile` and added a new `functions/Dockerfile-testing` just for this purpose.

File: pupil-spa/src/app/login-failure/login-failure.component.html (hunk 1 above)
Comment (danwilliams) [reply pupil-spa/src/app/login-failure/login-failure.component.html]: Nope, I was trying to troubleshoot why the pupil spa wasn’t connecting to the API from inside it’s container. Good spot! Removed.

File: admin/docker-start.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/docker-start.sh]: ah yes, good point 👍

File: docker-compose.pupil-test.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply docker-compose.pupil-test.yml]: those 2 are completely separate items...
`APPINSIGHTS_INSTRUMENTATIONKEY` - a single token value for authenticating to app insights
`AZURE_STORAGE_CONNECTION_STRING` - connection string to azure storage

**Comment by GuyHarwood** [comment]
@danwilliams also - please add documentation on how to set this off.

File: README.md [hunk 6]
```diff
@@ -32,8 +32,16 @@ Stand up the database containers with: `docker-compose up`
 ### docker-compose.admin-test.yml and docker-compose.pupil-test.yml
 Runs SQL Server, MongoDB, all web applications, and the pupil and admin test suites as separate containers.
 This compose file depends on `docker-compose.yml` and should not be run directly.
+
 To stand up the tests, app and db containers run either `./run-compose-admin-test-suite.sh` or `./run-compose-pupil-test-suite.sh`
 
+Both of these scripts require the following values defined in `/admin/.env`:
+```
+APPINSIGHTS_INSTRUMENTATIONKEY='foo-bar'
```
Comment (GuyHarwood) [inline README.md]: how come app insights is no longer optional, am i overlooking something?

Comment (danwilliams) [reply README.md]: In my notes, `AZURE_STORAGE_CONNECTION_STRING` and `APPINSIGHTS_INSTRUMENTATIONKEY` values are the same. So I originally had them both coming from the same env var. When you pointed out that they can be different values, I split them out. If `APPINSIGHTS_INSTRUMENTATIONKEY` isn’t required for testing, I can take it out.

**Comment by activemq** [comment]
@danwilliams is it possible to run the admin tests and the pupil test simultaneously? Would there be an issue with the container names conflicting? Would be a nice to have

**Comment by danwilliams** [comment]
@activemq They both rely on the `sqldb` and `redis` containers in `docker-compose.yml`. You’d have to create a copy of that which calls them `sqldb_2` and `redis_2`. Then use one for the admin tests and the other for the pupil tests.

The main issue is the machine you’re trying to do this on being able to handle 12 docker containers being built and run at the same time 🙂

**Comment by activemq** [comment]
@GuyHarwood FYI Admin tests having some issues ie. chromedriver crashing and memory trying to resolve these before approving

**Comment by activemq** [comment]
@GuyHarwood pushed the final changes to get the suite running within docker. There are tests failures but as discussed, will be tackling these in a separate pbi

**Comment by GuyHarwood** [comment]
Looks good, just one minor comment 🏆

**Comment by GuyHarwood** [comment]
@activemq see travis output - it looks like changes have been made to compose files that travis depends upon.  roll back the changes, and make separate copies of the files - one for travis, one for the test run, or find a way to share them for less maintenance overhead

**Pull request merged**